azurerm/resource_arm_monitor_diagnostics.go

```diff
@@ -0,0 +1,309 @@
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+
+			"event_hub_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"event_hub_authorization_rule_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"metric_settings": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MinItems: 1,
+				MaxItems: 16,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"category": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"retention_days": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"time_grain": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"log_settings": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MinItems: 1,
+				MaxItems: 16,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"category": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"retention_days": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmMonitorDiagnosticsCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing arguments for Azure ARM Diagnostic Settings.")
+
+	name := d.Get("name").(string)
+	resourceId := d.Get("resource_id").(string)
+	storageAccountId := d.Get("storage_account_id").(string)
+	eventHubName := d.Get("event_hub_name").(string)
+	eventHubAuthorizationRuleId := d.Get("event_hub_authorization_rule_id").(string)
+	workspaceId := d.Get("workspace_id").(string)
+	metricSettings := d.Get("metric_settings").(*schema.Set)
+	logSettings := d.Get("log_settings").(*schema.Set)
+
+	diagnosticSettings := &insights.DiagnosticSettings{}
+
+	if metricSettings != nil {
+		diagnosticSettings.Metrics = expandMetricsConfiguration(metricSettings)
+	}
+
+	if logSettings != nil {
+		diagnosticSettings.Logs = expandLogConfiguration(logSettings)
+	}
+
+	if len(storageAccountId) > 0 {
+		diagnosticSettings.StorageAccountID = &storageAccountId
+	}
+
+	if len(workspaceId) > 0 {
+		diagnosticSettings.WorkspaceID = &workspaceId
+	}
+
+	if len(eventHubAuthorizationRuleId) > 0 && len(eventHubName) > 0 {
+		diagnosticSettings.EventHubAuthorizationRuleID = &eventHubAuthorizationRuleId
+		diagnosticSettings.EventHubName = &eventHubName
+	}
+
+	_, err := client.CreateOrUpdate(
+		ctx,
+		resourceId,
+		insights.DiagnosticSettingsResource{
+			Name:               &name,
+			DiagnosticSettings: diagnosticSettings,
+		},
+		name)
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(ctx, resourceId, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Diagnostic Settings")
```

Comment:
can we make this error message clearer to match the other resources? e.g. `fmt.Errorf("Cannot read ID for Monitor Diagnostics %q", name)`